### PR TITLE
Fixes #33417 - Hide version from the Login page

### PR DESCRIPTION
**Pull request by avitova**
Work in progress

### CODE REVIEW

**Review by lzap — CHANGES_REQUESTED** [review]
Thanks for the patch, this would not work with upgrades.

Can you figure out some seach and replace mechanism framework for Settings and use "Version $VERSION" default value? It can be then search and replaced in the runtime.

**Comment by ezr-ondrej** [comment]
ok to test

**Comment by lzap** [comment]
@laviro can you help us resolving JS linter issues? Thanks

File: app/models/setting/general.rb
Comment (ezr-ondrej) [inline app/models/setting/general.rb]: It would be better to have a constant
```suggestion
      set('login_text', N_("Text to be shown in the login-page footer"), "Version $VERSION", N_('Login page footer text. Use $VERSION to be replace by current version.')),
```

File: app/models/setting/general.rb
Comment (Ron-Lavi) [inline app/models/setting/general.rb]: In this case, I would suggest adding a setting to display the version on the login page if someone wants to, which will be set to `false` by default.
later, in https://github.com/theforeman/foreman/blob/develop/app/helpers/login_helper.rb#L5 we can pass the version to the login page if that setting is set to `true`

File: webpack/assets/javascripts/react_app/components/LoginPage/LoginPage.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/LoginPage/LoginPage.js]: you can run `npm run lint -- --fix` to automatically fix the lining issue.

File: webpack/assets/javascripts/react_app/components/LoginPage/LoginPage.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/LoginPage/LoginPage.js:32]: instead of removing it, I would suggest adding a new setting that sets whether to display the version or not, and then in https://github.com/theforeman/foreman/blob/develop/app/helpers/login_helper.rb#L5 we can pass it or not based on that setting.

**Review by Ron-Lavi — COMMENTED** [review]
Thanks @avitova!

File: webpack/assets/javascripts/react_app/components/LoginPage/LoginPage.js
Comment (lzap) [reply webpack/assets/javascripts/react_app/components/LoginPage/LoginPage.js:32]: Apologies, there was a long discussion on the security list about how to approach this and folks did not like a new setting. So we decided either to remove the version completely, or move the text into the footer text which was a bit more flexible since QA do require version to be present for some tests.

File: app/models/setting.rb
Comment (lzap) [inline app/models/setting.rb]: It is a not good practice to modify arguments, just use `gsub` which will return a new string.

Comment (lzap) [reply app/models/setting.rb]: Also this method name could be called `replace_keywords` as we can add more later.

File: app/models/setting.rb
Comment (lzap) [inline app/models/setting.rb]: Only replace keywords for some names:

```
KEYWORD_REPLACEMENTS = [:login_text, :instance_title]

def ....
  value = replace_keyword(value) if KEYWORD_REPLACEMENTS.include?(name)
  ...
  value
end
```

Something like this I think.

**Review by lzap — CHANGES_REQUESTED** [review]
Tests are still failing.

File: webpack/assets/javascripts/react_app/components/LoginPage/LoginPage.js
Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/LoginPage/LoginPage.js:32]: Oh I see, it's still feels a bit strange to display `version $VERSION` in case the admins didn't replace the variable.
Adding @marisvirik for another opinion,
Maybe we can test if the $VERSION var has changed and only if it did, we should display it, does it makes sense?

**Comment by lzap** [comment]
@avitova please rebase and resolve conflicts

@laviro we need your help with JS tests failures

Thank you!

**Review by Ron-Lavi — COMMENTED** [review]
Great work @avitova !
looks like you solved also the JS test errors,
can you also remove the version which is passed from:
https://github.com/theforeman/foreman/blob/develop/app/helpers/login_helper.rb#L5

and also from: https://github.com/theforeman/foreman/blob/3347fa49d500964f0209122d8d36c920d1feafcc/app/helpers/external_logout_helper.rb#L4

is there another place that where we want to show the version? e.g.: about page
EDIT: never mind, I see it's already displayed there: https://github.com/theforeman/foreman/blob/69379b51a33344de93836950565ecbfae0c4df23/app/views/about/index.html.erb#L141

**Comment by avitova** [comment]
> Great work @avitova ! looks like you solved also the JS test errors, can you also remove the version which is passed from: https://github.com/theforeman/foreman/blob/develop/app/helpers/login_helper.rb#L5
> 
> and also from:
> 
> https://github.com/theforeman/foreman/blob/3347fa49d500964f0209122d8d36c920d1feafcc/app/helpers/external_logout_helper.rb#L4
> 
> is there another place that where we want to show the version? e.g.: about page EDIT: never mind, I see it's already displayed there:
> 
> https://github.com/theforeman/foreman/blob/69379b51a33344de93836950565ecbfae0c4df23/app/views/about/index.html.erb#L141

@laviro Sure, I will change those files as file. But it seems like it does not work now. As I just discovered, the file app/models/setting/general.rb was deleted 12 days ago as a part of Fixes #32893.

**Comment by tbrisker** [comment]
@avitova looks like something went wrong in the rebase, it seems that your commit got squashed with the message from my last commit

**Comment by avitova** [comment]
@tbrisker Hopefully resolved. Thanks, I saw there was an extra file included but didn't figure out the reason.

File: app/models/setting.rb
Comment (lzap) [inline app/models/setting.rb]: Probably check for string, looks cleaner.

**Review by lzap — CHANGES_REQUESTED** [review]
Can you add a test into `test/models/setting_test.rb`

**Review by ekohl — COMMENTED** [review]
I don't think replacement in the model is correct. For verification, I think this workflow must remain possible:
* User goes to settings page
* User sees `My Login $VERSION`

I think now the settings page will show the interpolated value of `$VERSION` instead of the raw `$VERSION`. The same thing probably affects the REST API.

To solve that I think you still need to modify the login and logout helpers and replace the caption there.

**Comment by lzap** [comment]
> To solve that I think you still need to modify the login and logout helpers and replace the caption there.

Sounds like you insist, @avitova can you do that then? Let's make it a special case then, drop the replacing in the accessor. This is fine.

File: app/models/setting.rb
Comment (ares) [reply app/models/setting.rb]: just to be clear, that means `.is_a?(String)` instead of using `respond_to?(:gsub)`

File: webpack/assets/javascripts/react_app/components/LoginPage/LoginPage.js
Comment (ares) [reply webpack/assets/javascripts/react_app/components/LoginPage/LoginPage.js:32]: FTR `$VERSION` keyword should never be displayed, we always replace it with the current version of Foreman. Admin can decide though to hide it and set the setting value to an empty string.

**Review by ares — COMMENTED** [review]
@ekohl I don't think you're right. I actually tested this and the settings shows the $VERSION string as well as API (tested through Hammer), while on login page it's correctly replaced. It's caused by the fact, we only override the `Setting.[...]` but our API uses `Setting.find(...).value`.

However I agree with you that this feels like a hack and it works thanks to unintentional code flow. And I agree we should only replace the $VERSION keyword in the presentation layer where we know, we actually want to replace it, essentially on the place we pass the data to the login page component. Otherwise it will bite us in the future.

File: app/models/setting.rb
Comment (Ron-Lavi) [reply app/models/setting.rb]: How should we handle version upgrades, would the user need to update the Setting each time?

Maybe for simplicity the setting should just set whether to display the version on Login page or not.

As for QE tests that expect the version, can we pass an ENV variable also to decide about displaying it or not to make life easier?

Comment (Ron-Lavi) [reply app/models/setting.rb]: discussed it with @lzap and @avitova offline and seem that it would work on upgrades,
so you can ignore my previous comment :)

File: app/helpers/login_helper.rb
Comment (lzap) [inline app/helpers/login_helper.rb]: Please move this into the Setting model class.

File: app/helpers/login_helper.rb
Comment (lzap) [inline app/helpers/login_helper.rb]: Drop this test please, it is not necessary.

**Review by lzap — CHANGES_REQUESTED** [review]
Few nitpicks, but this is nice and clean.

**Review by Ron-Lavi — COMMENTED** [review]
Nice work @avitova ! tested and the replacement works as expected 👍🏼 
can you also omit the `version`  in code, prop and propTypes as it won't be passed anymore: 

https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/LoginPage/LoginPage.js#L32

and also in:
https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/ExternalLogout/ExternalLogout.js#L26

File: webpack/assets/javascripts/react_app/components/LoginPage/LoginPage.js
Comment (lzap) [reply webpack/assets/javascripts/react_app/components/LoginPage/LoginPage.js:32]: You need to move the i18n of the "Version" word from JS here into Ruby. Problem is that we initialize default value during seed I think, if user changes language it will not be translated. I think we need to live with that.

**Review by Ron-Lavi — COMMENTED** [review]
almost there!

https://github.com/theforeman/foreman/runs/4351016298?check_suite_focus=true#step:8:2327
Complaints snapshot were updated after the version elements are gone,
can you run `npm test -- -u` to update them?

**Comment by avitova** [comment]
> almost there!
> 
> https://github.com/theforeman/foreman/runs/4351016298?check_suite_focus=true#step:8:2327 Complaints snapshot were updated after the version elements are gone, can you run `npm test -- -u` to update them?

Thank you, I was not sure how to handle this issue.

File: app/registries/foreman/settings/general.rb
Comment (ezr-ondrej) [inline app/registries/foreman/settings/general.rb]: Shall we have the default here to what used to be there on the page?
`default: 'Version $VERSION'`?

**Review by ezr-ondrej — COMMENTED** [review]
I'm quite happy what we ended up doing, I believe that we were talking about defaulting to showing the version tho, but I'm not sure if that didn't change in the process - I'm sorry I only peeked time to time :)
Apart of that it looks good to me :+1:

**Review by Ron-Lavi — DISMISSED** [review]
Thanks @avitova ! great work 👍🏼 
I rebased it for you and reverted snapshot changes that were not related to this PR

**Comment by Ron-Lavi** [comment]
@lzap you requested changes earlier, if everything seems good can you merge please?

File: app/registries/foreman/settings/general.rb
Comment (Ron-Lavi) [reply app/registries/foreman/settings/general.rb]: sorry, missed this comment.
iiuc I think the idea is to hide the version by default but to let the user the option to display it

File: app/models/setting.rb
Comment (lzap) [inline app/models/setting.rb:265]: I find this method a bit unsafe, when it's called with something else than string, then it returns `nil`. Can you change it to return actually the unchanged argument if it does not respond to `gsub`?

File: app/registries/foreman/settings/general.rb
Comment (lzap) [reply app/registries/foreman/settings/general.rb]: Yeah but Red Hat QA wants to keep the old text to be present. And that is "Version $VERSION" so I agree it should be the default value. Do not forget to use gettext: `N_("Version $VERSION")`

**Review by lzap — CHANGES_REQUESTED** [review]
Few comments, please squash into a single commit so it is ready for merge.

**Comment by avitova** [comment]
> Few comments, please squash into a single commit so it is ready for merge.

I did remove validation for responding to gsub, though some test did not really make it. I will try to change the method so that it meets the needs.

**Comment by lzap** [comment]
@Ron-Lavi can you tell waht is wrong with JS tests? Thanks! LGTM from my perspective.

File: app/registries/foreman/settings/general.rb
Comment (lzap) [inline app/registries/foreman/settings/general.rb]: Maybe a nitpick, it is more translator-friendly to do this: `N_("Version") + " $VERSION",`

**Comment by Ron-Lavi** [comment]
@avitova can you rebase and then run `rm -rf node_modules/ package-lock.json` & `npm install` ? then `npm test -- -u` to update the failing snapshots?

**Review by Ron-Lavi — APPROVED** [review]
Tests are now passing, and it seems to work well 👍🏼 thanks @avitova!
any other comments before merging? @lzap @ezr-ondrej @ekohl

**Review by lzap — APPROVED** [review]
Thanks please do the honour.

**Comment by Ron-Lavi** [comment]
Thanks @avitova @lzap !! 👍🏼

**Comment by ares** [comment]
Nice work, definitely worth demoing :-)

**Pull request merged**
